invenio_drafts_resources/records/systemfields/parent.py

```diff
@@ -33,23 +32,49 @@ def __init__(self, parent_record_cls, key=None, create=False,
 
     def create(self, record):
         """Method to create the parent record for the record."""
-        # Create a parent record
-        parent = self._parent_record_cls.create({})
-        # This bumps revision id to 2 for the first record. Commit needed
-        # because of the system fields defined on the parent record (e.g. pid).
-        parent.commit()
-
-        # Set using the __set__() method
-        setattr(record, self.attr_name, parent)
+        parent = getattr(record, self.attr_name)
+        if parent is None:
+            # Create a parent record
+            parent = record.parent_record_cls.create({})
+            # This bumps revision id to 2 for the first record. Commit needed
+            # because of the system fields defined on the parent record (e.g.
+            # pid).
+            parent.commit()
 
+            # Set using the __set__() method
+            setattr(record, self.attr_name, parent)
         return parent
 
     def delete(self, record, force=False):
-        """Method to delete the parent record."""
+        """Method to delete the parent record.
+
+        The parent record will only be hard deleted if this is the last
+        record/draft with links to it.
+        """
+        # A *record* is configured to never delete the parent
+        # automatically.
+        # A *draft* is configured to delete the parent on hard delete but not
+        # on soft delete.
+        #
+        # A draft may be hard-deleted in two cases: 1) a new record 2) a new
+        # record version. In the case 1, only a draft and the parent record
+        # exists. In case 2, one or more drafts/records may exists with a
+        # foreign key to the parent record.
+        #
+        # The logic implemented here, thus on a hard delete tries to delete the
+        # the parent record and relies on the database integrity constraints
+        # to prevent the parent record from being deleted in case more
+        # drafts/records exists.
         parent = getattr(record, self.attr_name)
         if parent:
             if force and self._hard_delete:
-                parent.delete(force=True)
+                try:
+                    with db.session.begin_nested():
```

Comment:
Comment: If an exception is detected, we rollback to the save point (otherwise the entire transaction needs to be rolled back). See also https://invenio-db.readthedocs.io/en/latest/session_management.html#what-happens-with-exceptions-in-nested-transactions